src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/Compression/WebSocketInflater.cs

```diff
@@ -0,0 +1,319 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
+using static System.IO.Compression.ZLibNative;
+
+namespace System.Net.WebSockets.Compression
+{
+    /// <summary>
+    /// Provides a wrapper around the ZLib decompression API.
+    /// </summary>
+    internal sealed class WebSocketInflater : IDisposable
+    {
+        internal const int FlushMarkerLength = 4;
+        internal static ReadOnlySpan<byte> FlushMarker => new byte[] { 0x00, 0x00, 0xFF, 0xFF };
+
+        private ZLibStreamHandle? _stream;
+        private readonly int _windowBits;
+        private readonly bool _persisted;
+
+        /// <summary>
+        /// There is no way of knowing, when decoding data, if the underlying deflater
+        /// has flushed all outstanding data to consumer other than to provide a buffer
+        /// and see whether any bytes are written. There are cases when the consumers
+        /// provide a buffer exactly the size of the uncompressed data and in this case
+        /// to avoid requiring another read we will use this field.
+        /// </summary>
+        private byte? _remainingByte;
+
+        /// <summary>
+        /// When the inflater is persisted we need to manually append the flush marker
+        /// before finishing the decoding.
+        /// </summary>
+        private bool _needsFlushMarker;
+
+        private byte[]? _buffer;
+
+        /// <summary>
+        /// The position for the next unconsumed byte in the inflate buffer.
+        /// </summary>
+        private int _position;
+
+        /// <summary>
+        /// How many unconsumed bytes are left in the inflate buffer.
+        /// </summary>
+        private int _available;
+
+        internal WebSocketInflater(int windowBits, bool persisted)
+        {
+            Debug.Assert(windowBits >= 9 && windowBits <= 15);
+
+            // We use negative window bits to instruct deflater to expect raw deflate data
+            _windowBits = -windowBits;
+            _persisted = persisted;
+        }
+
+        public bool Finished { get; private set; } = true;
+
+        public Memory<byte> Memory => _buffer;
+
+        public Span<byte> Span => _buffer;
+
+        public void Dispose()
+        {
+            _stream?.Dispose();
+            ReleaseBuffer();
+        }
+
+        public void Initialize(long payloadLength)
+        {
+            Debug.Assert(_available == 0);
+            Debug.Assert(_buffer is null);
+
+            // Do not try to rent anythin above 1MB because the array pool
+            // will not pool the buffer but allocate it.
+            _buffer = ArrayPool<byte>.Shared.Rent((int)Math.Min(payloadLength, 1_000_000));
```

Comment:
We should be ignoring the incoming payload length and instead starting with a tiny buffer that grows (doubles) up to some maximum.

Currently, a client can force this logic to allocate big buffers per websocket.

We also should delay this allocation for as long as possible.

Reply:
I see your point, but resizing here is not an option at the moment, due to constraints in the implementation to keep the original receive method, so we don't introduce unwanted regressions for the uncompressed messages. However I agree with you that 1MB is a bit too much here, and we can restrict the maximum inflate buffer to be for example 64KB?

About delaying it further - this is as late as it can be - the websocket needs the buffer to start receiving.

Reply:
Seems like this would need to be configurable (deflate stream defaults to 8K). I would start with something based on the input buffer size as that is basically all the user has to state their intent.

Reply:
@davidfowl Is this better - https://github.com/dotnet/runtime/pull/49304/files#diff-0edaec6e3a3fe21900f67d3293d6b93ebd1e1507fcc789e3de7f5623ed7e055bR77-R92?